driver.go

```diff
@@ -81,7 +81,8 @@ func (d MySQLDriver) Open(dsn string) (driver.Conn, error) {
 		}
 	}
 
-	mc.buf = newBuffer(mc.netConn)
+	// Make the buffer get the context from mysqlConn.
+	mc.buf = newBufferContext(mc.netConn, &ContextWrapper{mc: mc})
```

Comment:
There are two reasons why I didn't use `netConn.Close()` when `<- ctx.Done()`. One is even if `ctx.Done()` happens and no matter it calls `netConn.Close()` or `netConn.SetDeadline()`, `mysqlConn.pingImpl()` may still send nil error to [line 65](https://github.com/go-sql-driver/mysql/pull/572/files#diff-7cd962f1d3aee9adf93d5c17b49f4530R65) before these two functions interrupt it. In this situation, `netConn` doesn't have to be closed and it's safe to reuse this connection. BTW, `mysqlConn.Ping()` checks the received `err` at [line 72](https://github.com/go-sql-driver/mysql/pull/572/files#diff-7cd962f1d3aee9adf93d5c17b49f4530R72). If the `err` shows the connection timed out, it will returns ErrBadConn and let caller close it.

The another reason is about the life cycle of `mysqlConn`. I think It is created and managed in `database/sql/sql.go`. It should be destroyed in the same level/place. Actually it does in [`DB.putConn()`](https://github.com/golang/go/blob/release-branch.go1.8/src/database/sql/sql.go#L608) if `Pinger.Ping()` returns ErrBadConn. Let `DB.putConn()` does what its duty is and `mysqlConn.Ping()` should not make `DB.putConn()` close it again.